### PR TITLE
Do not hardcode "lib/rpm" as the installation path for default config…

### PR DESCRIPTION
**Pull request by kanavin**
…uration and macros.

This can be named in various different ways, especially in cross-compilation
environments, so let's take it from the build setup.

### CODE REVIEW

File: macros.in [hunk 1]
```diff
@@ -942,7 +942,7 @@ package or when debugging this package.\
 %_sharedstatedir	%{_prefix}/com
 %_localstatedir		%{_prefix}/var
 %_lib			lib
-%_libdir		%{_exec_prefix}/%{_lib}
+%_libdir		@libdir@
```
Comment (ignatenkobrain) [inline macros.in]: +1

File: rpm.am [hunk 2]
```diff
@@ -1,10 +1,10 @@
 # Internal binaries
 ## HACK: It probably should be $(libexecdir)/rpm or $(libdir)/rpm
-rpmlibexecdir = $(prefix)/lib/rpm
+rpmlibexecdir = $(libdir)/rpm
```
Comment (ignatenkobrain) [inline rpm.am]: +1

File: configure.ac [hunk 3]
```diff
@@ -1005,7 +1005,7 @@ else
     usrprefix=$prefix
 fi
 
-RPMCONFIGDIR="`echo ${usrprefix}/lib/rpm`"
+RPMCONFIGDIR="`echo ${libdir}/rpm`"
```
Comment (ignatenkobrain) [inline configure.ac]: We probably should really use libexecdir

File: rpm.am (hunk 2 above)
Comment (Conan-Kudo) [inline rpm.am]: This should be `$(libexecdir)` here too.

File: macros.in (hunk 1 above)
Comment (Conan-Kudo) [inline macros.in]: I think the reason it's like this is to support the weird multiarch schemes that some distributions use... 🤷‍♂️

**Comment by kanavin** [comment]
How about now?

**Comment by Conan-Kudo** [comment]
@kanavin I think the only issue is the change to `%_libdir`. Is that really necessary? It's set up that way so that it can be flexible. I'd rather see this change reverted.

**Comment by kanavin** [comment]
Previously it was set this way:
```
%_lib			lib
%_libdir		%{_exec_prefix}/%{_lib}
```
So basically 'lib' was hardcoded via %_lib. If you show me how to avoid that in a different way than my patch, I can do it.

**Comment by Conan-Kudo** [comment]
@kanavin What are you trying to accomplish?

**Comment by kanavin** [comment]
@Conan-Kudo In Yocto project the libdir can be lib, lib32, lib64, or anything else. You cannot assume anything about it, and when building the software, the value for it is typically passed to ./configure as a command line switch.

On the other hand, %_libdir macro hardcodes the value to be /usr/lib, and then that macro is used all over the place in rpm. So I am removing that hardcoding, and setting the macro dynamically at build time from the build environment settings.

**Comment by pmatilai** [comment]
Much of the stuff in /usr/lib/rpm could live in %{_datadir}/rpm but there are also ELF executables there that can not, so the stuff would have to be split into two different places. And there are a LOT of hardcoded references to /usr/lib/rpm out there in packages, all of which would break for no particularly good reason.

I think everybody agrees that /usr/lib/rpm is not exactly optimal place for this stuff but in reality the cure is far worse than the disease.

**Comment by Conan-Kudo** [comment]
@kanavin The values in macros.in are overridden by the generated platform macros: https://github.com/rpm-software-management/rpm/blob/master/platform.in

As far as I know, as long as you aren't purging the platforms, this doesn't matter.

**Comment by kanavin** [comment]
@pmatilai But they are split into two different places:

 # Internal binaries
rpmlibexecdir = $(libexecdir)/rpm

# Host independent config files
rpmconfigdir = $(datadir)/rpm

I don't mind where that stuff lives (it could stay in libdir), as long as the 'lib' part is not hardcoded, like it is now:
rpmlibexecdir = $(prefix)/lib/rpm
rpmconfigdir = $(prefix)/lib/rpm

**Comment by Conan-Kudo** [comment]
@pmatilai We could at least move things to `$(libexecdir)/rpm` and for Fedora have a symlink for the old location to point to the new one. That'd keep things consistent, right?

**Comment by ignatenkobrain** [comment]
My personal preference is to keep $prefix/lib/rpm hardcoded. We do this in systemd, casync and a lot of other tools.

**Comment by kanavin** [comment]
@Conan-Kudo in Yocto the generated platform macros are not used, as there is a hardcoded list of those platforms, which is not flexible enough for us. So we directly override settings from the default macros file at runtime.

However, for some reason _libdir is not overriden in the same way, and is instead configured at build time - so you have a point, I should change that.

**Comment by pmatilai** [comment]
Splitting into two introduces two dozen new headaches that didn't exist before because even rpm itself expects all the cruft in a single place.

Sorry but NAK, I see nothing here that would even remotely justify breaking 20+ years worth of tradition and introducing new complications in the process.

**Comment by kanavin** [comment]
@pmatilai Okay, splitting things into two places is bad. How about not hardcoding the 'lib' part? Should I open a separate PR for it?

**Pull request closed without merging**